base/reflection.jl

```diff
@@ -233,6 +234,55 @@ a concrete type that can have instances, or the empty type `Union{}`.
 isleaftype(t::ANY) = (@_pure_meta; (isa(t, DataType) && t.isleaftype) || t === Union{})
 
 """
+    Base.isabstract(T)
+
+Determine whether `T` was declared as an abstract type (i.e. using the
+`abstract` keyword).
+"""
+function isabstract(t::ANY)
+   @_pure_meta
+   t = unwrap_unionall(t)
+   isa(t,DataType) && t.abstract
+end
+
+"""
+    Base.parameter_upper_bound(t::UnionAll, idx)
+
+Determine the upper bound of a type parameter in the underlying type. E.g.:
+```jldoctest
+julia> immutable Foo{T<:AbstractFloat, N}
+           x::Tuple{T, N}
+       end
+
+julia> Base.parameter_upper_bound(Foo, 1)
+AbstractFloat
+
+julia> Base.parameter_upper_bound(Foo, 2)
+Any
+```
+"""
+function parameter_upper_bound(t::UnionAll, idx)
+   @_pure_meta
+   rewrap_unionall(unwrap_unionall(t).parameters[idx], t)
+end
+
+"""
+    Base.unparameterized_type(t)
+
+Given a (partially) specialized type, return the unspecialized type, e.g.:
+```jldoctest
+julia> immutable Foo{T}
+           x::T
+       end
+
+julia> Base.unparameterized_type(Foo{Float64})
+Foo
+```
+"""
+unparameterized_type(t::DataType) = t.name.wrapper
+unparameterized_type(t::UnionAll) = unparameterized_type(unwrap_unionall(t))
```

Comment:
That transform isn't meaningful for an arbitrary type `T`. It's actually not really even valid for a non-arbitrary type `T` (for example, `f(D::Type{Dict}, K, V) = D{K, V}` cannot be inferred, leading to this needed correction: https://github.com/JuliaLang/julia/pull/18457/commits/210d4ec450676c44b6011542f468c66020d82bb3#diff-d4736682004c1996e5d9b1c5c33c8c8e).

In the only cases where it is valid (e.g. you passed a type literal to `basetype`), you could just write that same type literal without the call to `basetype`. And the resulting code would also be less error-prone if someone replaced the original type with a type-alias (in addition to being shorter, by virtue of not containing no-op function calls).